alfalfa_worker/lib/job.py

```diff
@@ -0,0 +1,265 @@
+# from alfalfa_jobs.run import Run
+from alfalfa_worker.lib.run import Run, RunStatus
+from alfalfa_worker.lib.sim_type import SimType
+
+
+def message(func):
+    """Decorator for methods which are triggered by messages.
+    Overriding a function that has a decorator in the base class (like 'stop') will need to be decorated again."""
+    setattr(func, 'message_handler', True)
+    return func
+
+
+def with_run(func):
+    """Decorator for methods which require a run.
+    Will throw an error if called before a run is associated with the job."""
+
+    def wrap(self, *args, **kwargs):
+        if self.run is None:
+            raise JobExceptionInvalidRun("Job has no associated run")
+        return func(self, *args, **kwargs)
+    return wrap
+
+
+class JobMetaclass(type):
+    """The purpose of the metaclass is to wrap the __init__ of the subclass.
+    This allows the run_manager argument to be handled the same way
+    in every job. These arguments are removed before calling the old __init__ function.
+    This wrapper also does other setup functions like initializing redis and setting
+    up message handlers for annotated functions."""
+    # Wrap Subclass __init__. Called once for every class of Job
+    def __new__(cls, name, bases, cls_dicts):
+        if '__init__' in cls_dicts.keys():
+            __old_init__ = cls_dicts['__init__']
+        else:
+            __old_init__ = None
+
+        def __new_init__(self, *args, **kwargs):
+            if hasattr(self, '_status') and self._status.value >= JobStatus.INITIALIZING.value:
+                if __old_init__:
+                    __old_init__(self, *args, **kwargs)
+                return
+            self.run_manager = kwargs.get('run_manager')
+            del kwargs['run_manager']
+            self._status = JobStatus.INITIALIZING
+
+            # Variables
+            self.run = None
+
+            # Redis
+            self.redis = Redis(host=os.environ['REDIS_HOST'])
+            self.redis_pubsub = self.redis.pubsub(ignore_subscribe_messages=True)
+            logging.basicConfig(level=os.environ.get("LOGLEVEL", "INFO"))
+            self.logger = logging.getLogger(self.__class__.__name__)
+
+            # Message Handlers
+            self._message_handlers = {}
+            for attr_name in dir(self):
+                attr = getattr(self, attr_name)
+                if hasattr(attr, 'message_handler'):
+                    self._message_handlers[attr_name] = attr
+            if __old_init__:
+                __old_init__(self, *args, **kwargs)
+            self.set_job_status(JobStatus.INITIALIZED)
+        cls_dicts['__init__'] = __new_init__
+        klazz = super().__new__(cls, name, bases, cls_dicts)
+
+        # This adds the job class to a list of all of the loaded jobs
+        if len(bases) == 0:
+            klazz.jobs = []
+        if len(bases) > 0:
+            Job.jobs.append(klazz)
+        return klazz
+
+
+class Job(metaclass=JobMetaclass):
```

Comment:
we should break these out into their own classes if possible.

Reply:
I'm not sure what you mean by this